sklearn/semi_supervised/_self_training.py

```diff
@@ -0,0 +1,261 @@
+    David Yarowsky. 1995. Unsupervised word sense disambiguation rivaling
+    supervised methods. In Proceedings of the 33rd annual meeting on
+    Association for Computational Linguistics (ACL '95). Association for
+    Computational Linguistics, Stroudsburg, PA, USA, 189-196. DOI:
+    https://doi.org/10.3115/981658.981684
+    """
+    _estimator_type = "classifier"
+
+    def __init__(self,
+                 base_classifier,
+                 threshold=0.75,
+                 max_iter=10):
+        self.base_classifier = base_classifier
+        self.threshold = threshold
+        self.max_iter = max_iter
+
+    def fit(self, X, y):
+        """
+        Fits this ``SelfTrainingClassifier`` to a dataset.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+            array representing the data
+
+        y : array-like, shape = (n_samples,)
+            array representing the labels. Unlabeled samples should have the
+            label -1.
+
+        Returns
+        -------
+        self : object
+            returns an instance of self.
+        """
+        X, y = check_X_y(X, y)
+        _validate_estimator(self.base_classifier)
+        self.base_classifier_ = clone(self.base_classifier)
+
+        if self.max_iter is not None and self.max_iter < 0:
+            msg = "max_iter must be >= 0 or None, got {}".format(self.max_iter)
+            raise ValueError(msg)
+
+        if not 0 <= self.threshold < 1:
+            msg = "threshold must be in [0,1), got {}".format(self.threshold)
+            raise ValueError(msg)
+
+        has_label = y != -1
+
+        if np.all(has_label):
+            warnings.warn("y contains no unlabeled samples", UserWarning)
+
+        self.y_labels_ = np.copy(y)
+        self.y_labeled_iter_ = np.full_like(y, -1)
+        self.y_labeled_iter_[has_label] = 0
+
+        self.n_iter_ = 0
+        while not np.all(has_label) and (self.max_iter is None or
+                                         self.n_iter_ < self.max_iter):
+            self.n_iter_ += 1
+            self.base_classifier_.fit(
+                X[safe_mask(X, has_label)],
+                self.y_labels_[safe_mask(self.y_labels_, has_label)])
+
+            # Predict on the unlabeled samples
+            prob = self.base_classifier_.predict_proba(
+                X[safe_mask(X, ~has_label)])
+            pred = self.base_classifier_.classes_[np.argmax(prob, axis=1)]
+            max_proba = np.max(prob, axis=1)
+
+            # Select samples where confidence is above the threshold
+            confident_labels_mask = max_proba > self.threshold
+
+            new_labels_idx = np.flatnonzero(~has_label)[confident_labels_mask]
+
+            # Add newly labeled confident predictions to the dataset
+            self.y_labels_[new_labels_idx] = pred[confident_labels_mask]
+            has_label[new_labels_idx] = True
+            self.y_labeled_iter_[new_labels_idx] = self.n_iter_
+
```

Comment:
If not all samples have labels at this point we should raise a ConvergenceWarning

Reply:
We discussed this [here](https://github.com/scikit-learn/scikit-learn/pull/11682#discussion_r218807594), and came to the conclusion that we shouldn't do it.

Reply:
For example with MNIST, which has 1797 datapoints:
![figure_1](https://user-images.githubusercontent.com/22137236/48767983-458ff500-ecb8-11e8-9ced-b2670640ae8a.png)